presto-main/src/main/java/com/facebook/presto/spiller/TempStorageSingleStreamSpiller.java

```diff
@@ -168,6 +168,10 @@ private void writePages(Iterator<Page> pageIterator)
                     });
         }
 
+        // Flush remaining buffered pages, if any
+        if (!bufferedPages.isEmpty()) {
+            flushBufferedPages();
```

Comment:
This should work great if enough data is spilled at once (e.g.: the entire hash table when memory is revoked). However in many cases this interface is also used to spill one page at a time (e.g.: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java#L368). Flushing one page at a time will result in a small block creation problem.

The buffer size is a constant value (probably ~24MB in production settings). A single instance of `SingleStreamSpiller` is rather unlikely to create a problem. In the problematic query how many `SingleStreamSpiller` instances are active? I wonder if we can try to re-adjust the settings to decrease the buffer size and decrease the concurrency?

Reply:
Yes. The buffer size is configured to be 24MB in size. In the problematic query, we have 8 instances of SpillableHashAggregationBuilder. Each of these holds around 3.5K instances of TempStorageSingleStreamSpiller, each of these instances can hold <24MB of pages in memory. From the heap dump of this query, I see each instance of  TempStorageSingleStreamSpiller holding around 0.2MB of pages on Avg (Very few instances hold 10MB and a large number of instances are holding ~100KB of pages). 

The current spilling logic creates a new instance of TempStorageSingleStreamSpiller for each time memory revoke is triggered. The query is probably triggering memory revoke very frequently and that is creating large instances of TempStorageSingleStreamSpiller, each one holding very less number of pages in memory. This eventually fills up the memory and triggers an OOM.

Since, TempStorageSingleStreamSpiller is holding very less data per instances, reducing the buffer size wont help and I don't think we can control the concurrency of `TempStorageSingleStreamSpiller` instances with the given logic.

Reply:
This issue is now fixed in https://github.com/prestodb/presto/pull/16537/ using a different approach.